### ISSUE TITLE
clang crashes on valid code at -Os and above: Assertion `idx < size()' failed 

### ISSUE THREAD

**Issue by zhendongsu** [llvm/llvm-project#62665]
It appears to be a recent regression.

Compiler Explorer: https://godbolt.org/z/TTaPb979W

```console
% clangtk -v
clang version 17.0.0 (https://github.com/llvm/llvm-project.git 33b69b97565cf19f8e11173fd851e56ce07656f0)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /local/suz-local/opfuzz/bin
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/11
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/12
Selected GCC installation: /usr/lib/gcc/x86_64-linux-gnu/12
Candidate multilib: .;@m64
Selected multilib: .;@m64
Found CUDA installation: /usr/local/cuda, version 11.8
% 
% clangtk -Os small.c
clang-17: /local/suz-local/software/clangbuild/llvm-project/llvm/include/llvm/ADT/SmallVector.h:294: T& llvm::SmallVectorTemplateCommon<T, <template-parameter-1-2> >::operator[](llvm::SmallVectorTemplateCommon<T, <template-parameter-1-2> >::size_type) [with T = int; <template-parameter-1-2> = void; llvm::SmallVectorTemplateCommon<T, <template-parameter-1-2> >::reference = int&; llvm::SmallVectorTemplateCommon<T, <template-parameter-1-2> >::size_type = long unsigned int]: Assertion `idx < size()' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: /local/home/suz/suz-local/software/local/clang-trunk/bin/clang-17 -cc1 -triple x86_64-unknown-linux-gnu -emit-obj -dumpdir a- -disable-free -clear-ast-before-backend -main-file-name small.c -mrelocation-model pic -pic-level 2 -pic-is-pie -mframe-pointer=none -fmath-errno -ffp-contract=on -fno-rounding-math -mconstructor-aliases -funwind-tables=2 -target-cpu x86-64 -tune-cpu generic -debugger-tuning=gdb -fcoverage-compilation-dir=/local/suz-local/software/emitesting/bugs/20230511-clangtk-m64-O3-Wall-Wextra-pipe-fPIC-build-141720/delta -resource-dir /local/home/suz/suz-local/software/local/clang-trunk/lib/clang/17 -I /usr/local/include/csmith -I /local/suz-local/software/local/include -internal-isystem /local/home/suz/suz-local/software/local/clang-trunk/lib/clang/17/include -internal-isystem /usr/local/include -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/12/../../../../x86_64-linux-gnu/include -internal-externc-isystem /usr/include/x86_64-linux-gnu -internal-externc-isystem /include -internal-externc-isystem /usr/include -Os -fdebug-compilation-dir=/local/suz-local/software/emitesting/bugs/20230511-clangtk-m64-O3-Wall-Wextra-pipe-fPIC-build-141720/delta -ferror-limit 19 -fgnuc-version=4.2.1 -fcolor-diagnostics -vectorize-loops -vectorize-slp -faddrsig -D__GCC_HAVE_DWARF2_CFI_ASM=1 -o /tmp/small-d75a82.o -x c small.c
1.	<eof> parser at end of file
2.	Optimizer
 #0 0x000055810bf017b0 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-17+0x3a907b0)
 #1 0x000055810beff094 SignalHandler(int) Signals.cpp:0:0
 #2 0x00007f9becb68520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #3 0x00007f9becbbca7c pthread_kill (/lib/x86_64-linux-gnu/libc.so.6+0x96a7c)
 #4 0x00007f9becb68476 gsignal (/lib/x86_64-linux-gnu/libc.so.6+0x42476)
 #5 0x00007f9becb4e7f3 abort (/lib/x86_64-linux-gnu/libc.so.6+0x287f3)
 #6 0x00007f9becb4e71b (/lib/x86_64-linux-gnu/libc.so.6+0x2871b)
 #7 0x00007f9becb5fe96 (/lib/x86_64-linux-gnu/libc.so.6+0x39e96)
 #8 0x000055810d574914 addMask(llvm::SmallVectorImpl<int>&, llvm::ArrayRef<int>, bool) (.constprop.0) SLPVectorizer.cpp:0:0
 #9 0x000055810d5bae8f llvm::slpvectorizer::BoUpSLP::reorderTopToBottom() (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-17+0x5149e8f)
#10 0x000055810d5efcbe llvm::SLPVectorizerPass::tryToVectorizeList(llvm::ArrayRef<llvm::Value*>, llvm::slpvectorizer::BoUpSLP&, bool) (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-17+0x517ecbe)
#11 0x000055810d5f3c4c bool tryToVectorizeSequence<llvm::Value>(llvm::SmallVectorImpl<llvm::Value*>&, llvm::function_ref<bool (llvm::Value*, llvm::Value*)>, llvm::function_ref<bool (llvm::Value*, llvm::Value*)>, llvm::function_ref<bool (llvm::ArrayRef<llvm::Value*>, bool)>, bool, llvm::slpvectorizer::BoUpSLP&) (.constprop.0) SLPVectorizer.cpp:0:0
#12 0x000055810d5f4fb5 llvm::SLPVectorizerPass::vectorizeChainsInBlock(llvm::BasicBlock*, llvm::slpvectorizer::BoUpSLP&) (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-17+0x5183fb5)
#13 0x000055810d5f9261 llvm::SLPVectorizerPass::runImpl(llvm::Function&, llvm::ScalarEvolution*, llvm::TargetTransformInfo*, llvm::TargetLibraryInfo*, llvm::AAResults*, llvm::LoopInfo*, llvm::DominatorTree*, llvm::AssumptionCache*, llvm::DemandedBits*, llvm::OptimizationRemarkEmitter*) (.part.0) SLPVectorizer.cpp:0:0
#14 0x000055810d5f9c12 llvm::SLPVectorizerPass::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-17+0x5188c12)
#15 0x000055810d15a076 llvm::detail::PassModel<llvm::Function, llvm::SLPVectorizerPass, llvm::PreservedAnalyses, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-17+0x4ce9076)
#16 0x000055810975839a llvm::detail::PassModel<llvm::Function, llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>, llvm::PreservedAnalyses, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-17+0x12e739a)
#17 0x000055810b8ae683 llvm::ModuleToFunctionPassAdaptor::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-17+0x343d683)
#18 0x000055810973d926 llvm::detail::PassModel<llvm::Module, llvm::ModuleToFunctionPassAdaptor, llvm::PreservedAnalyses, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-17+0x12cc926)
#19 0x000055810b8ac441 llvm::PassManager<llvm::Module, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-17+0x343b441)
#20 0x000055810c160380 (anonymous namespace)::EmitAssemblyHelper::RunOptimizationPipeline(clang::BackendAction, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>&, std::unique_ptr<llvm::ToolOutputFile, std::default_delete<llvm::ToolOutputFile>>&) BackendUtil.cpp:0:0
#21 0x000055810c1639b2 clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::StringRef, llvm::Module*, clang::BackendAction, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>) (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-17+0x3cf29b2)
#22 0x000055810d0f6769 clang::BackendConsumer::HandleTranslationUnit(clang::ASTContext&) (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-17+0x4c85769)
#23 0x000055810e05cb79 clang::ParseAST(clang::Sema&, bool, bool) (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-17+0x5bebb79)
#24 0x000055810c95beb9 clang::FrontendAction::Execute() (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-17+0x44eaeb9)
#25 0x000055810c8ddb1e clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-17+0x446cb1e)
#26 0x000055810ca26446 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-17+0x45b5446)
#27 0x00005581093b0be9 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-17+0xf3fbe9)
#28 0x00005581093ac8b3 ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&, llvm::ToolContext const&) driver.cpp:0:0
#29 0x00005581093adf50 clang_main(int, char**, llvm::ToolContext const&) (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-17+0xf3cf50)
#30 0x00005581092c2c33 main (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-17+0xe51c33)
#31 0x00007f9becb4fd90 (/lib/x86_64-linux-gnu/libc.so.6+0x29d90)
#32 0x00007f9becb4fe40 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x29e40)
#33 0x00005581093a70f5 _start (/local/home/suz/suz-local/software/local/clang-trunk/bin/clang-17+0xf360f5)
clangtk: error: unable to execute command: Aborted
clangtk: error: clang frontend command failed due to signal (use -v to see invocation)
clang version 17.0.0 (https://github.com/llvm/llvm-project.git 33b69b97565cf19f8e11173fd851e56ce07656f0)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /local/suz-local/opfuzz/bin
clangtk: note: diagnostic msg: 
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clangtk: note: diagnostic msg: /tmp/small-c39741.c
clangtk: note: diagnostic msg: /tmp/small-c39741.sh
clangtk: note: diagnostic msg: 

********************
% 
% cat small.c
int printf(const char *, ...);
int a, b, c, d, e, f, g, i, j, k, l, m, n, r;
int main() {
  int p = 0, *o = &p;
  if (d)
    p--;
  if (g)
    goto q;
  if (b) {
    int x = 0, s = a;
    if (x)
    t:
      if (p)
        o = &j;
    r = s;
    if (d)
      k = p;
    if (i)
      goto u;
    int v = g;
    if (p)
      while (p++)
        ;
  u:
    if (d)
      printf("%d", p);
    while (m)
      if (p)
        printf("0");
    if (e)
      goto ab;
    if (v)
      n++;
  }
  *o = 0;
  int q = r;
 q:
  l = p;
 ab:
  c = a ? q : 0;
  if (a)
    goto t;
  return 0;
}
```

**Comment by fhahn** [comment]
Looks like this started with the commit below. cc @alexey-bataev 


```
commit 1ce4b26a21a0b7a263f9ddc14827d79a87fb2542
Author: Alexey Bataev <a.bataev@outlook.com>
Date:   Thu Apr 13 11:28:54 2023 -0700

    [SLP]Add final resize to ShuffleCostEstimator::finalize member function and basic add member functions.

    Implemented the reshuffling in finalize member function + add basic
    support for add member functions, used during vector build.

    Part of D110978

    Differential Revision: https://reviews.llvm.org/D148279
```

**Comment by alexey-bataev** [comment]
Fixed in ae5ff3ca0cd923a9fb2b46701641acab3dd15e4b
